src/Parser/TemplateRenderer.cs

```diff
@@ -0,0 +1,68 @@
+﻿using System;
+using System.Text;
+using Parser.Parts;
+
+namespace Parser
+{
+    public static class TemplateRenderer
+    {
+
+        public static string Render(this Template template, IFormatProvider formatProvider, object[] parameters)
+        {
+            var sb = new StringBuilder(template.Value.Length + 64 * template.Holes.Length);
+            int pos = 0;
+            int holeIndex = 0;
+            foreach (var literal in template.Literals)
+            {
+                sb.Append(template.Value, pos, literal.Print);
+                pos += literal.Print;
+                if (literal.Skip == 0)
+                {
+                    pos++;
+                }
+                else
+                {
+                    pos += literal.Skip;
+                    if (template.IsPositional)
+                    {
+                        RenderHolePositional(sb, template.Holes[holeIndex++], formatProvider, parameters);
+                    }
+                    else
+                    {
+                        RenderHole(sb, template.Holes[holeIndex], formatProvider, parameters[holeIndex++]);
+                    }
+                }
+            }
+            return sb.ToString();
+        }
+
+        private static void RenderHolePositional(StringBuilder sb, Hole hole, IFormatProvider formatProvider, object[] parameters)
```

Comment:
Nitpicking but I'm not sure this method adds a lot, as the code could be directly above.
In particular, it contains as much logic as the other if-branch, which did not receive its own `RenderHoleNamed` method.